### PR TITLE
Fix exception on missing manifest start_url

### LINKED ISSUE #557: Lighthouse crashes on manifests without a URL
```
Runtime error encountered: TypeError: Parameter "url" must be a string, not undefined
    at Url.parse (url.js:87:11)
    at urlParse (url.js:81:5)
    at Url.resolve (url.js:635:29)
    at Object.urlResolve [as resolve] (url.js:631:40)
    at Function.audit (/Users/surma/src/github.com/GoogleChrome/lighthouse/lighthouse-core/audits/cache-start-url.js:58:26)
    at Promise.all.config.audits.map.audit (/Users/surma/src/github.com/GoogleChrome/lighthouse/lighthouse-core/runner.js:70:38)
    at Array.map (native)
    at run.then.artifacts (/Users/surma/src/github.com/GoogleChrome/lighthouse/lighthouse-core/runner.js:67:61)
TypeError: Parameter "url" must be a string, not undefined
    at Url.parse (url.js:87:11)
    at urlParse (url.js:81:5)
    at Url.resolve (url.js:635:29)
    at Object.urlResolve [as resolve] (url.js:631:40)
    at Function.audit (/Users/surma/src/github.com/GoogleChrome/lighthouse/lighthouse-core/audits/cache-start-url.js:58:26)
    at Promise.all.config.audits.map.audit (/Users/surma/src/github.com/GoogleChrome/lighthouse/lighthouse-core/runner.js:70:38)
    at Array.map (native)
    at run.then.artifacts (/Users/surma/src/github.com/GoogleChrome/lighthouse/lighthouse-core/runner.js:67:61)
```

### PR DESCRIPTION
**Pull request by paulirish**
This may fix #557, however I was unable to reproduce the exception, even trying a manifest with "start_url: undefined` and missing the property entirely.

@surma can you try this out?

### CODE REVIEW

**Comment by surma** [comment]
Yes, this fixes it for me :)

File: lighthouse-core/audits/cache-start-url.js [hunk 1]
```diff
@@ -39,20 +39,23 @@ class CacheStartUrl extends Audit {
...
+    const manifest = artifacts.Manifest && artifacts.Manifest.value;
+    const cacheContents = artifacts.CacheContents;
+    const baseURL = artifacts.URL;
 
-    if (!(artifacts.Manifest &&
-          artifacts.Manifest.value &&
-          artifacts.Manifest.value.start_url &&
-          Array.isArray(artifacts.CacheContents) &&
-          artifacts.URL)) {
+    if (!(manifest && manifest.start_url && manifest.start_url.raw)) {
```
Comment (brendankenny) [inline lighthouse-core/audits/cache-start-url.js]: can just be `if (!(manifest && manifest.start_url.raw)) {` (if there is a manifest, there is always a `manifest.start_url`)

Comment (surma) [inline lighthouse-core/audits/cache-start-url.js]: Isn’t that the exact bug we are trying to catch here? That there’s cases where `start_url` is in fact NOT defined?

File: lighthouse-core/test/audits/cache-start-url.js [hunk 2]
```diff
@@ -48,6 +48,18 @@ describe('Cache: start_url audit', () => {
...
     const inputs = {
       Manifest: {
+
+      }
+    };
+
+    return assert.equal(Audit.audit(inputs).rawValue, false);
+  });
+
+  // Need to disable camelcase check for dealing with short_name.
```
Comment (brendankenny) [inline lighthouse-core/test/audits/cache-start-url.js]: s/short_name/start_url

File: lighthouse-core/test/audits/cache-start-url.js [hunk 3]
```diff
@@ -48,6 +48,18 @@ describe('Cache: start_url audit', () => {
...
+  /* eslint-disable camelcase */
+  it('fails when a manifest contains no start_url', () => {
```
Comment (brendankenny) [inline lighthouse-core/test/audits/cache-start-url.js]: `'...contains null start_url'` maybe?

File: lighthouse-core/audits/cache-start-url.js (hunk 1 above)
Comment (brendankenny) [inline lighthouse-core/audits/cache-start-url.js]: `manifest.start_url` is always defined. It's [`manifest.start_url.raw`](https://github.com/GoogleChrome/lighthouse/blob/6663f6b0bab11f0971fe390ddbb2a167496d9c57/lighthouse-core/audits/cache-start-url.js#L58) that was causing the error before

Comment (brendankenny) [inline lighthouse-core/audits/cache-start-url.js]: should also use `manifest.start_url.value` not `raw` here

**Comment by brendankenny** [comment]
1) @surma is there a test that can be added that will trigger the error on current master?
2) @paulirish while you're in cache-start-url.js, can you add a sentence or two file overview? It's not really clear which cache it's talking about, why you'd care, etc

**Comment by paulirish** [comment]
thx for the review. will update.

**Comment by surma** [comment]
@paulirish @brendankenny 

This test in `test/audits/cache-start-url.js` fails for me on master:

```js
  it('doesn\'t throw when no start_url is in manifest', () => {
    return assert.equal(Audit.audit({
      Manifest: {
        value: {
          name: {},
          start_url: {} // eslint-disable-line camelcase
        }
      },
      CacheContents,
      URL
    }).rawValue, true);
  });
```

**Comment by brendankenny** [comment]
oh, just noticed that these tests are passing in manifest artifacts directly, but the manifest artifacts being passed in aren't possible outputs of the manifest parser :)

We should be doing the old
```
  // Need to disable camelcase check for dealing with start_url.
  /* eslint-disable camelcase */
  it('fails when a manifest contains no start_url', () => {
    const inputs = {
      Manifest: manifestParser({})
    };

    return assert.equal(Audit.audit(inputs).rawValue, false);
  });
```

**Comment by paulirish** [comment]
> oh, just noticed that these tests are passing in manifest artifacts directly, but the manifest artifacts being passed in aren't possible outputs of the manifest parser :)

yeahhhh I'm a big +1 on using manifest parser within like all of our manifest-y audit tests. Its dangerous to try to mock the artifact object directly.

File: lighthouse-core/test/audits/cache-start-url.js [hunk 4]
```diff
@@ -43,18 +37,30 @@ describe('Cache: start_url audit', () => {
...
         start_url: null
       }
     };
-
     return assert.equal(Audit.audit(inputs).rawValue, false);
   });
 
+  it('fails when a manifest contains no start_url', () => {
+    const inputs = {
+      Manifest: manifestParser({})
```
Comment (brendankenny) [inline lighthouse-core/test/audits/cache-start-url.js]: Will you switch the two tests above to use the parser too? They may be redundant with this one at that point, not sure.

**Comment by brendankenny** [comment]
LGTM though brownie points if you fix old tests :)

**Comment by wardpeet** [comment]
would this also fix #539 ? (which I couldn't reproduce)

**Pull request merged**